settings.js

```diff
@@ -0,0 +1,256 @@
+
+        const pageData = pages[page];
+        pageData[id] = {
+            position: GLib.Variant.new_int32(position),
+        }
+
+        addedItems.add(itemId);
+        index++;
+    }
+
+    // Switch to the next page
+    index = itemsPerPage * Math.ceil(index / itemsPerPage);
+
+    // Retrieve apps inside folders
+    const appsInsideFolders = new Set();
+    folderSettings.get_strv('folder-children').forEach(id => {
+        const path = `${folderSettings.path}folders/${id}/`;
+        const folder = new Gio.Settings({
+            schema_id: 'org.gnome.desktop.app-folders.folder',
+            path,
+        });
+
+        const folderApps = _getAppsInsideFolder(installedApps, folder);
+        folderApps.forEach(appId => appsInsideFolders.add(appId));
+    });
+
+    for (const app of installedApps) {
+        const appId = app.get_id();
+
+        if (!app.should_show() ||
+            addedItems.has(appId) ||
+            appsInsideFolders.has(appId) ||
+            appId.startsWith(EOS_LINK_PREFIX))
+            continue;
+
+        const page = Math.floor(index / itemsPerPage);
+        const position = index % itemsPerPage;
+
+        if (page === pages.length)
+            pages.push({});
+
+        const pageData = pages[page];
+        pageData[appId] = {
+            position: GLib.Variant.new_int32(position),
+        }
+
+        index++;
+    }
+
+    // Store the new page layout
+    const packedPages = [];
+
+    // Pack the icon properties as a GVariant
+    for (const page of pages) {
+        const pageData = {};
+        for (const [appId, properties] of Object.entries(page))
+            pageData[appId] = new GLib.Variant('a{sv}', properties);
+        packedPages.push(pageData);
+    }
+
+    const variant = new GLib.Variant('aa{sv}', packedPages);
+    global.settings.set_value('app-picker-layout', variant);
+}
+
+function applyMigration() {
+    const migrationSettings = _getMigrationSettings();
+    const extensionSettings = ExtensionUtils.getSettings();
+
+    const migrationVTable = [
+        _migrateToV1,
+    ];
+
+    let migrationVersion = extensionSettings.get_uint('migration-version');
+    if (migrationVersion > migrationVTable.length - 1)
+        return;
+
+    while (migrationVersion < CURRENT_VERSION) {
+        const migrationFunc = migrationVTable[migrationVersion];
+
+        migrationFunc(migrationSettings, extensionSettings);
+        migrationVersion++;
+
+        extensionSettings.set_uint('migration-version', migrationVersion);
+    }
+}
+
+function migrate() {
+    const parentalControlsManager = ParentalControlsManager.getDefault();
+
+    if (parentalControlsManager.initialized) {
+        applyMigration();
+        return;
+    }
+
+    // Wait for parental controls to be ready, otherwise we
+    // get an empty set of icons
+    const id =
+        parentalControlsManager.connect('app-filter-changed', () => {
+            if (!parentalControlsManager.initialized)
```

Comment:
I see a small issue here in case the extension gets disabled before the parental controls get initialized - in which case we would not disconnect this handler - not sure if worth updating though.

Another point is what happens if parental controls takes too long to initialize - do we show the grid without migrating and then update it afterwards? Should we hide the grid until migration runs?

Reply:
I'm developing this in my old Endless One and this entire migration path runs before the Shell even shows up. I'm positive that if it works in such a slow machine, there's no reason to assume that it'll misbehave anywhere else.

But, it is a good point that migration should be cancelled if the extension is disabled.

Reply:
I pondered about it more, and I believe it's actually better to make the migration a non-optional, non-cancelable part of the extension. It is easier for us to not have to deal with the edge cases like "what if the migration ran halfway" or "should it cancel migration if extension is disabled" or etc.

Reply:
Yes, that is what I meant by "not sure if worth..." - I think this is a really edge case where someone enables/disables the extension before the parental controls is initialized. Lets go with it as is and if needed (unlikely) we fix it later on.